src/riak_kv_qry.erl

```diff
@@ -281,22 +281,60 @@ count_to_position(Col, [_ | Rest], Pos) ->
                        {ok, query_tabular_result()} | {error, term()}.
 do_select(SQL, ?DDL{table = BucketType} = DDL) ->
     Mod = riak_ql_ddl:make_module_name(BucketType),
-    MaxSubQueries =
-        app_helper:get_env(riak_kv, timeseries_query_max_quanta_span),
 
     case riak_ql_ddl:is_query_valid(Mod, DDL, riak_kv_ts_util:sql_record_to_tuple(SQL)) of
         true ->
-            case riak_kv_qry_compiler:compile(DDL, SQL, MaxSubQueries) of
+            case riak_kv_qry_compiler:compile(DDL, SQL) of
                 {error,_} = Error ->
                     Error;
                 {ok, SubQueries} ->
-                    maybe_await_query_results(
-                      riak_kv_qry_queue:put_on_queue(self(), SubQueries, DDL))
+                    FullCycle =
+                        fun(QBufRef) ->
+                            maybe_await_query_results(
+                              riak_kv_qry_queue:put_on_queue(self(), SubQueries, DDL, QBufRef))
+                        end,
+                    case maybe_create_query_buffer(SQL, length(SubQueries), DDL, []) of
+                        {error, Reason} ->
+                            %% query buffers are non-optional for
+                            %% ORDER BY queries
+                            {error, {qbuf_create_error, Reason}};
+                        {ok, undefined} ->
+                            %% query buffer path not advisable (query
+                            %% has no ORDER BY)
+                            FullCycle(undefined);
+                        {ok, {new, QBufRef}} ->
+                            %% query is eligible, and new buffer is ready
+                            FullCycle(QBufRef);
+                        {ok, {existing, QBufRef}} ->
+                            %% query is eligible AND it is a follow-up
+                            %% query: yay results!
+                            riak_kv_qry_buffers:fetch_limit(
+                              QBufRef,
+                              SQL?SQL_SELECT.'LIMIT', SQL?SQL_SELECT.'OFFSET')
+                    end
             end;
         {false, Errors} ->
             {error, {invalid_query, format_query_syntax_errors(Errors)}}
     end.
 
+
+-spec maybe_create_query_buffer(?SQL_SELECT{}, pos_integer(), ?DDL{}, proplists:proplist()) ->
```

Comment:
Yes, `maybe_create_query_buffer` has the intended effect for eligible queries (and is a noop for others).

Re `undefined` for the `QBufRef` it returns, it's a way to avoid two extra indent levels. An alternative would be to have a case statement differentiating on `?SQL_SELECT{'ORDER BY' = DefinedOrNot}`, where `create_query_buffer` would do its work unconditionally. And, the `FullCycle` and downstream functions have been taught to handle `undefined` for `QBufRef`.